zebra-chain/src/primitives/redpallas/scalar_mul.rs

```diff
@@ -0,0 +1,207 @@
+// -*- mode: rust; -*-
+//
+// This file is part of redpallas.
+// Copyright (c) 2019-2021 Zcash Foundation
+// Copyright (c) 2017-2021 isis agora lovecruft, Henry de Valence
+// See LICENSE for licensing information.
+//
+// Authors:
+// - isis agora lovecruft <isis@patternsinthevoid.net>
+// - Henry de Valence <hdevalence@hdevalence.ca>
+// - Deirdre Connolly <deirdre@zfnd.org>
+
+use std::{borrow::Borrow, fmt::Debug};
+
+use group::Group;
+use halo2::{arithmetic::FieldExt, pasta::pallas};
+
+pub trait NonAdjacentForm {
+    fn non_adjacent_form(&self, w: usize) -> [i8; 256];
+}
+
+/// A trait for variable-time multiscalar multiplication without precomputation.
+pub trait VartimeMultiscalarMul {
+    /// The type of point being multiplied, e.g., `AffinePoint`.
+    type Point;
+
+    /// Given an iterator of public scalars and an iterator of
+    /// `Option`s of points, compute either `Some(Q)`, where
+    /// $$
+    /// Q = c\_1 P\_1 + \cdots + c\_n P\_n,
+    /// $$
+    /// if all points were `Some(P_i)`, or else return `None`.
+    fn optional_multiscalar_mul<I, J>(scalars: I, points: J) -> Option<Self::Point>
+    where
+        I: IntoIterator,
+        I::Item: Borrow<pallas::Scalar>,
+        J: IntoIterator<Item = Option<Self::Point>>;
+
+    /// Given an iterator of public scalars and an iterator of
+    /// public points, compute
+    /// $$
+    /// Q = c\_1 P\_1 + \cdots + c\_n P\_n,
+    /// $$
+    /// using variable-time operations.
+    ///
+    /// It is an error to call this function with two iterators of different lengths.
+    fn vartime_multiscalar_mul<I, J>(scalars: I, points: J) -> Self::Point
+    where
+        I: IntoIterator,
+        I::Item: Borrow<pallas::Scalar>,
+        J: IntoIterator,
+        J::Item: Borrow<Self::Point>,
+        Self::Point: Clone,
+    {
+        Self::optional_multiscalar_mul(
+            scalars,
+            points.into_iter().map(|p| Some(p.borrow().clone())),
+        )
+        .unwrap()
+    }
+}
+
+impl NonAdjacentForm for pallas::Scalar {
+    /// Compute a width-\\(w\\) "Non-Adjacent Form" of this scalar.
+    ///
+    /// Thanks to curve25519-dalek
```

Comment:
Would it be useful to point to the specific commit/release from curve25519-dalek this was based on, to make it easier to compare/update the code if needed? e.g. https://github.com/dalek-cryptography/curve25519-dalek/blob/3e189820da03cc034f5fa143fc7b2ccb21fffa5e/src/scalar.rs#L907

Reply:
Done via d8ef0ee4